src/main/java/hudson/plugins/audit_trail/CredentialUsageListener.java

```diff
@@ -0,0 +1,65 @@
+package hudson.plugins.audit_trail;
+
+import com.cloudbees.plugins.credentials.Credentials;
+import com.cloudbees.plugins.credentials.CredentialsUseListener;
+import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import hudson.Extension;
+import hudson.model.AbstractModelObject;
+import hudson.model.Item;
+import hudson.model.Run;
+
+import javax.inject.Inject;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+@Extension
+public class CredentialUsageListener extends CredentialsUseListener {
+    @Inject
+    AuditTrailPlugin configuration;
+
+    @Override
+    public void onUse(Credentials c, AbstractModelObject obj) {
+        if (!configuration.shouldLogCredentialsUsage())
+            return;
+
+        StringBuilder builder = new StringBuilder(100);
+        String objName = obj.toString();
+        String objType = obj.getClass().toString();
```

Comment:
Not great. Pending https://issues.jenkins.io/browse/JENKINS-26091 you would likely need to checkcast the kind of object. Or change the listener interface to have one method per kind of model object: `Node`, `Item`, or `Run` it seems.

Reply:
Agreed - it is not optimal as I rely on the implementation of the toString method of the object. However, the output of Node and Run seemed sufficient:
Run:

> 'ExampleFolder/ExampleFreeStyle #2' (class hudson.model.FreeStyleBuild) used credentials 'ExampleUserNamePassword' (class com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl).

Node:

> 'hudson.slaves.DumbSlave[ExampleNode]' (class hudson.slaves.DumbSlave) used credentials 'ExampleUserNamePassword' (class com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl).

(Did not yet find a usecase for item)

Your comment in the other PR (https://github.com/jenkinsci/credentials-plugin/pull/295#discussion_r835784622) made me simplify this even further (see latest commit), removing all type checks.

It would be possible to check for types to have cleaner output for Node (only name), but for example for Runs, I would probably only want the current output anyways. I think the advantage of this approach might be the greater flexibility and simplicity. I could of course implement a checkcast for Node, Item and Run and in all other cases fall back to this approach. What are your thoughts on this?

Reply:
The current output seems fine if you only care about human inspection. If you want the audit trail to be machine-readable it would be better to either downcast `ModelObject`s, or just adjust the listener interface to include overloaded methods with more specific types.

Reply:
Fair point. I tested the output for some other objects (Folder, as seen in tests) and atleast for the usecase I have in mind it fits the requirements. The audit logs are handled by humans for the most part I think.